dataplane/src/packet.rs

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Open Network Fabric Authors
+
+use net::buffer::PacketBufferMut;
+use net::eth::EthError;
+use net::headers::{AbstractHeaders, AbstractHeadersMut, Headers, TryHeaders, TryHeadersMut};
+use net::parse::{DeParse, DeParseError, Parse, ParseError};
+use std::cmp::Ordering;
+use std::num::NonZero;
+
+#[derive(Debug)]
+pub struct Packet<Buf: PacketBufferMut> {
+    headers: Headers,
+    /// The total number of bytes _originally_ consumed when parsing this packet
+    /// Mutations to `packet` can cause the re-serialized size of the packet to grow or shrink.
+    consumed: NonZero<u16>,
+    pub mbuf: Buf, // TODO: find a way to make this private
+}
+#[derive(Debug, thiserror::Error)]
+pub struct InvalidPacket<Buf: PacketBufferMut> {
+    #[allow(unused)]
+    mbuf: Buf,
+    #[source]
+    error: ParseError<EthError>,
+}
+
+impl<Buf: PacketBufferMut> Packet<Buf> {
+    pub fn new(mbuf: Buf) -> Result<Packet<Buf>, InvalidPacket<Buf>> {
+        let (headers, consumed) = match Headers::parse(mbuf.as_ref()) {
+            Ok((packet, consumed)) => (packet, consumed),
+            Err(error) => {
+                return Err(InvalidPacket { mbuf, error });
+            }
+        };
+        Ok(Packet {
+            headers,
+            consumed,
+            mbuf,
+        })
+    }
+
+    pub(crate) fn reserialize(self) -> Buf {
+        // TODO: prove that these unreachable statements are optimized out
+        // The `unreachable` statements in the first block should be easily optimized out, but best
+        // to confirm.
+        let needed = self.headers.size();
+        let mut mbuf = self.mbuf;
+        let mut mbuf = match needed.cmp(&self.consumed) {
+            Ordering::Equal => mbuf,
+            Ordering::Less => {
+                let prepend = needed.get() - self.consumed.get();
+                match mbuf.prepend(prepend) {
+                    Ok(_) => {}
+                    Err(e) => unreachable!("configuration error: {:?}", e),
+                };
+                mbuf
+            }
+            Ordering::Greater => {
+                let trim = self.consumed.get() - needed.get();
+                if trim > self.headers.size().get() {
+                    panic!("attempting to trim a nonsensical amount of data: {trim}");
+                }
+                match mbuf.trim_from_start(trim) {
+                    Ok(_) => {}
+                    Err(e) => unreachable!("configuration error: {:?}", e),
```

Comment:
Not blocking, but we will want to make an issue requiring us to confirm, at the management plane level, that we don't accept configurations which require more headroom than we actually get.

That might be a tricky chore tho.

Reply:
This is extra tricky if we can't count on having a specific nic.

We will need the dataplane to report available headroom in a given mbuf at startup and forward that the the management plane when it validates configs.

That complicates things because a config which is valid on one machine might be illegal on another network card.

It is likely worth a change to dpdk-sys to ensure we have 256 bytes of headroom just to decrease the likelihood of causing problems for customers.

That said, we will still _need_ the check regardless of how much headroom we configure.

Reply:
I think the better answer here is probably to allow prechaining mbufs in the Packet interface instead of raising this error.  It'll cost performance but not cause issues for the customer.